includes/gateways/paypal/includes/class-wc-gateway-paypal-request.php

```diff
@@ -91,12 +91,15 @@ public function get_request_url( $order, $sandbox = false ) {
 	 * @return string
 	 */
 	protected function limit_length( $string, $limit = 127 ) {
-		// As the output is to be used in http_build_query which applies URL encoding, the string needs to be
-		// cut as if it was URL-encoded, but returned non-encoded (it will be encoded by http_build_query later).
-		$url_encoded_str = rawurlencode( $string );
-
-		if ( strlen( $url_encoded_str ) > $limit ) {
-			$string = rawurldecode( substr( $url_encoded_str, 0, $limit - 3 ) . '...' );
+		$str_limit = $limit - 3;
+		if ( function_exists( 'mb_strimwidth' ) ) {
+			if ( mb_strlen( $string ) > $limit ) {
+				$string = mb_strimwidth( $string, 0, $str_limit ) . '...';
+			}
+		} else {
+			if ( strlen( $string ) > $limit ) {
+				$string = substr( $string, 0, $str_limit ) . '...';
+			}
```

Comment:
This `rawurlencode` + `rawurldecode` were added to make sure that urlencoded string is shorter than 127 characters (based on https://developer.paypal.com/docs/classic/paypal-payments-standard/integration-guide/Appx_websitestandard_htmlvariables/?mark=night_phone_b#individual-items-variables). When looking at it now, I'm not sure if that was a correct reading of the spec.

It seems though, that PayPal does not care about the length that much anyway, so I'm not sure if it's an actual problem or not (the example string ends up as 410 characters when URL-encoded). The code previously cut off the string at byte 124 which could produce an invalid UTF8 character at the end of the string.

Side note: Funnily enough, it seems that item_name is processed by `wc_trim_string` first, which have an interesting side effect: it first checks if number of bytes in the string is larger than the limit (which it is, as the 77 UTF8 chars in the example are not from Latin alphabet), but then trims the string using multibyte char function, so instead of trimming the string, it just appends `...` at the end, so `Шампунь для увлажнения и питания волос Шампунь для увлажнения и питания волос` becomes `Шампунь для увлажнения и питания волос Шампунь для увлажнения и питания волос...`.

Long story short, this seems to work fine, not sure if we need to somehow limit the number of characters in other way, if that's what the PP docs says.

Reply:
I'm not sure they care either. In testing I went over the limit without it erring. I do think this keeps it down to the character limit AFTER decoding entities though so I think it's good enough...

Do you think the ... appending thing is something that needs to be looked at further before this PR is approved?

Reply:
No, that's just an unimportant detail, let's get this in.